File: src/lib/loanMath.ts
import type { LoanProduct, ScheduleInstalment } from '../types'

function addPeriod(date: Date, frequency: LoanProduct['repaymentFrequency']): Date {
  const d = new Date(date)
  switch (frequency) {
    case 'daily':
      d.setDate(d.getDate() + 1)
      return d
    case 'weekly':
      d.setDate(d.getDate() + 7)
      return d
    case 'fortnightly':
      d.setDate(d.getDate() + 14)
      return d
    case 'monthly':
      d.setMonth(d.getMonth() + 1)
      return d
  }
}

// Converts the product's per-period interest rate into a rate that applies once per instalment,
// based on how often interest is quoted (daily/weekly/monthly) vs how often instalments fall due.
function ratePerInstalment(product: LoanProduct): number {
  const periodDays: Record<LoanProduct['interestPeriod'], number> = {
    daily: 1,
    weekly: 7,
    monthly: 30,
  }
  const freqDays: Record<LoanProduct['repaymentFrequency'], number> = {
    daily: 1,
    weekly: 7,
    fortnightly: 14,
    monthly: 30,
  }
  const dailyRate = product.interestRate / 100 / periodDays[product.interestPeriod]
  return dailyRate * freqDays[product.repaymentFrequency]
}

export function totalFeeAmount(product: LoanProduct, principal: number, timing?: 'deducted' | 'added'): number {
  return product.fees
    .filter((f) => (timing ? f.timing === timing : true))
    .reduce((sum, f) => sum + (f.kind === 'percent' ? (principal * f.value) / 100 : f.value), 0)
}

export function generateSchedule(
  product: LoanProduct,
  principal: number,
  termInstalments: number,
  startDate: Date = new Date(),
): ScheduleInstalment[] {
  const rate = ratePerInstalment(product)
  const addedFees = totalFeeAmount(product, principal, 'added')
  const graceOnPrincipal = product.gracePeriodDays > 0 && (product.gracePeriodAppliesTo === 'principal' || product.gracePeriodAppliesTo === 'both')
  const graceOnInterest = product.gracePeriodDays > 0 && (product.gracePeriodAppliesTo === 'interest' || product.gracePeriodAppliesTo === 'both')

  const principalPeriods = graceOnPrincipal ? termInstalments - 1 : termInstalments
  const basePrincipalPortion = principal / Math.max(principalPeriods, 1)

  const schedule: ScheduleInstalment[] = []
  let balance = principal
  let dueDate = new Date(startDate)

  for (let period = 1; period <= termInstalments; period++) {
    dueDate = addPeriod(dueDate, product.repaymentFrequency)
    const isGracePeriod = period === 1 && product.gracePeriodDays > 0

    let interestDue = 0
    if (product.interestMethod === 'reducing') {
      interestDue = isGracePeriod && graceOnInterest ? 0 : balance * rate
    } else {
      const flatInterestPerPeriod = (principal * (product.interestRate / 100)) // per the quoted period, applied evenly
      interestDue = isGracePeriod && graceOnInterest ? 0 : flatInterestPerPeriod
    }

    const principalDue = isGracePeriod && graceOnPrincipal ? 0 : basePrincipalPortion
    const feesDue = period === 1 ? addedFees : 0

    balance = Math.max(balance - principalDue, 0)

    schedule.push({
      period,
      dueDate: dueDate.toISOString(),
      principalDue: round2(principalDue),
      interestDue: round2(interestDue),
      feesDue: round2(feesDue),
      penaltyDue: 0,
      totalDue: round2(principalDue + interestDue + feesDue),
      paidAmount: 0,
      balanceAfter: round2(balance),
      status: 'upcoming',
    })
  }

  return schedule
}

export function round2(n: number): number {
  return Math.round(n * 100) / 100
}

export function calculatePenalty(product: LoanProduct, overdueAmount: number, daysLate: number): number {
  if (daysLate <= 0) return 0
  const raw =
    product.penaltyKind === 'fixed' ? product.penaltyValue * daysLate : overdueAmount * (product.penaltyValue / 100) * daysLate
  return round2(Math.min(raw, product.penaltyCap))
}

export function classifyArrearsBand(daysLate: number): string {
  if (daysLate <= 0) return 'current'
  if (daysLate <= 7) return '1-7'
  if (daysLate <= 30) return '8-30'
  if (daysLate <= 60) return '31-60'
  if (daysLate <= 90) return '61-90'
  return 'over90'
}

export interface AllocationResult {
  schedule: ScheduleInstalment[]
  allocation: { penalty: number; fees: number; interest: number; principal: number }
  remainder: number
}

export function allocatePayment(
  product: LoanProduct,
  schedule: ScheduleInstalment[],
  amount: number,
): AllocationResult {
  const next = schedule.map((s) => ({ ...s }))
  const allocation = { penalty: 0, fees: 0, interest: 0, principal: 0 }
  let remaining = round2(amount)

  for (const inst of next) {
    if (remaining <= 0) break
    const instRemaining = round2(inst.totalDue - inst.paidAmount)
    if (instRemaining <= 0) continue

    const payForInst = round2(Math.min(instRemaining, remaining))
    const bucketDue: Record<'penalty' | 'fee' | 'interest' | 'principal', number> = {
      penalty: inst.penaltyDue,
      fee: inst.feesDue,
      interest: inst.interestDue,
      principal: inst.principalDue,
    }

    // Work out how much of each bucket earlier partial payments on this instalment already covered.
    let alreadyPaid = inst.paidAmount
    const bucketPaidSoFar: Record<string, number> = { penalty: 0, fee: 0, interest: 0, principal: 0 }
    for (const bucket of product.allocationOrder) {
      const take = Math.min(bucketDue[bucket], alreadyPaid)
      bucketPaidSoFar[bucket] = take
      alreadyPaid = round2(alreadyPaid - take)
    }

    let toApply = payForInst
    for (const bucket of product.allocationOrder) {
      if (toApply <= 0) break
      const capacity = round2(bucketDue[bucket] - bucketPaidSoFar[bucket])
      const take = Math.min(capacity, toApply)
      if (take <= 0) continue
      toApply = round2(toApply - take)
      if (bucket === 'fee') allocation.fees += take
      else if (bucket === 'penalty') allocation.penalty += take
      else if (bucket === 'interest') allocation.interest += take
      else allocation.principal += take
    }

    inst.paidAmount = round2(inst.paidAmount + payForInst)
    inst.status = inst.paidAmount >= inst.totalDue - 0.01 ? 'paid' : 'partial'
    remaining = round2(remaining - payForInst)
  }

  return { schedule: next, allocation, remainder: round2(remaining) }
}

File: backend/lms/services/loan_math.py
"""Loan scheduling, fee and repayment-allocation math.

A direct port of the frontend's src/lib/loanMath.ts, kept in lockstep with it:
the backend must compute the same numbers the frontend prototype does, and must
never trust a client-submitted schedule or allocation.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import date, datetime

from dateutil.relativedelta import relativedelta

PERIOD_DAYS = {"daily": 1, "weekly": 7, "monthly": 30}
FREQUENCY_DAYS = {"daily": 1, "weekly": 7, "fortnightly": 14, "monthly": 30}


def round2(value: float) -> float:
    return round(value + 1e-9, 2)


def _add_period(d: date, frequency: str) -> date:
    if frequency == "daily":
        return d + relativedelta(days=1)
    if frequency == "weekly":
        return d + relativedelta(weeks=1)
    if frequency == "fortnightly":
        return d + relativedelta(weeks=2)
    if frequency == "monthly":
        return d + relativedelta(months=1)
    raise ValueError(f"unknown repayment frequency: {frequency}")


def _rate_per_instalment(product) -> float:
    daily_rate = float(product.interest_rate) / 100 / PERIOD_DAYS[product.interest_period]
    return daily_rate * FREQUENCY_DAYS[product.repayment_frequency]


def _fees(product):
    fees = product.fees
    return fees.all() if hasattr(fees, "all") else fees


def total_fee_amount(product, principal: float, timing: str | None = None) -> float:
    total = 0.0
    for fee in _fees(product):
        if timing is not None and fee.timing != timing:
            continue
        if fee.kind == "percent":
            total += principal * float(fee.value) / 100
        else:
            total += float(fee.value)
    return total


@dataclass
class ScheduleRow:
    period: int
    due_date: date
    principal_due: float
    interest_due: float
    fees_due: float
    penalty_due: float
    total_due: float
    paid_amount: float = 0.0
    balance_after: float = 0.0
    status: str = "upcoming"


def generate_schedule(product, principal: float, term_instalments: int, start_date: datetime | date) -> list[ScheduleRow]:
    if isinstance(start_date, datetime):
        start_date = start_date.date()

    rate = _rate_per_instalment(product)
    added_fees = total_fee_amount(product, principal, "added")

    grace_on_principal = product.grace_period_days > 0 and product.grace_period_applies_to in ("principal", "both")
    grace_on_interest = product.grace_period_days > 0 and product.grace_period_applies_to in ("interest", "both")

    principal_periods = term_instalments - 1 if grace_on_principal else term_instalments
    base_principal_portion = principal / max(principal_periods, 1)

    schedule: list[ScheduleRow] = []
    balance = principal
    due_date = start_date

    for period in range(1, term_instalments + 1):
        due_date = _add_period(due_date, product.repayment_frequency)
        is_grace_period = period == 1 and product.grace_period_days > 0

        if product.interest_method == "reducing":
            interest_due = 0.0 if (is_grace_period and grace_on_interest) else balance * rate
        else:
            flat_interest_per_period = principal * (float(product.interest_rate) / 100)
            interest_due = 0.0 if (is_grace_period and grace_on_interest) else flat_interest_per_period

        principal_due = 0.0 if (is_grace_period and grace_on_principal) else base_principal_portion
        fees_due = added_fees if period == 1 else 0.0

        balance = max(balance - principal_due, 0.0)

        schedule.append(
            ScheduleRow(
                period=period,
                due_date=due_date,
                principal_due=round2(principal_due),
                interest_due=round2(interest_due),
                fees_due=round2(fees_due),
                penalty_due=0.0,
                total_due=round2(principal_due + interest_due + fees_due),
                paid_amount=0.0,
                balance_after=round2(balance),
                status="upcoming",
            )
        )

    return schedule


def calculate_penalty(product, overdue_amount: float, days_late: int) -> float:
    if days_late <= 0:
        return 0.0
    if product.penalty_kind == "fixed":
        raw = float(product.penalty_value) * days_late
    else:
        raw = overdue_amount * (float(product.penalty_value) / 100) * days_late
    return round2(min(raw, float(product.penalty_cap)))


@dataclass
class Allocation:
    penalty: float = 0.0
    fees: float = 0.0
    interest: float = 0.0
    principal: float = 0.0


@dataclass
class AllocationResult:
    schedule: list[ScheduleRow]
    allocation: Allocation
    remainder: float


def allocate_payment(product, schedule: list[ScheduleRow], amount: float) -> AllocationResult:
    """Applies a payment across a schedule in the product's configured bucket order,
    oldest unpaid instalment first. Mirrors allocatePayment in loanMath.ts exactly."""

    # `schedule` may be ScheduleRow copies or ORM instalments (DecimalField
    # columns yield Decimal), so every amount is coerced to float up front.
    next_schedule = [
        ScheduleRow(
            period=row.period,
            due_date=row.due_date,
            principal_due=float(row.principal_due),
            interest_due=float(row.interest_due),
            fees_due=float(row.fees_due),
            penalty_due=float(row.penalty_due),
            total_due=float(row.total_due),
            paid_amount=float(row.paid_amount),
            balance_after=float(row.balance_after),
            status=str(row.status),
        )
        for row in schedule
    ]
    allocation = Allocation()
    remaining = round2(amount)

    for inst in next_schedule:
        if remaining <= 0:
            break
        inst_remaining = round2(inst.total_due - inst.paid_amount)
        if inst_remaining <= 0:
            continue

        pay_for_inst = round2(min(inst_remaining, remaining))
        bucket_due = {
            "penalty": inst.penalty_due,
            "fee": inst.fees_due,
            "interest": inst.interest_due,
            "principal": inst.principal_due,
        }

        already_paid = inst.paid_amount
        bucket_paid_so_far = {"penalty": 0.0, "fee": 0.0, "interest": 0.0, "principal": 0.0}
        for bucket in product.allocation_order:
            take = min(bucket_due[bucket], already_paid)
            bucket_paid_so_far[bucket] = take
            already_paid = round2(already_paid - take)

        to_apply = pay_for_inst
        for bucket in product.allocation_order:
            if to_apply <= 0:
                break
            capacity = round2(bucket_due[bucket] - bucket_paid_so_far[bucket])
            take = min(capacity, to_apply)
            if take <= 0:
                continue
            to_apply = round2(to_apply - take)
            if bucket == "fee":
                allocation.fees += take
            elif bucket == "penalty":
                allocation.penalty += take
            elif bucket == "interest":
                allocation.interest += take
            else:
                allocation.principal += take

        inst.paid_amount = round2(inst.paid_amount + pay_for_inst)
        inst.status = "paid" if inst.paid_amount >= inst.total_due - 0.01 else "partial"
        remaining = round2(remaining - pay_for_inst)

    return AllocationResult(schedule=next_schedule, allocation=allocation, remainder=round2(remaining))
